src/Workspaces/Core/Portable/Workspace/Solution/ProjectIdMapping.cs

```diff
@@ -0,0 +1,116 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Frozen;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis;
+
+/// <summary>
+/// Mapping from ProjectId to values of type T. Set of possible ProjectID keys are determined at construction.
+/// Population of values are done through calls to GetOrAdd.
+/// </summary>
+/// <typeparam name="T"></typeparam>
+internal struct ProjectIdMapping<T>
+    where T : class
+{
+    public static ProjectIdMapping<T> Empty = new([], FrozenDictionary<ProjectId, int>.Empty);
+
+    /// <summary>
+    /// Mapping from ProjectId to index in _projects
+    /// </summary>
+    private readonly FrozenDictionary<ProjectId, int> _projectIdToIndex;
+
+    /// <summary>
+    /// Values are created on demand by calls to <see cref="GetOrAdd{TArg}(ProjectId?, Func{ProjectId, TArg, T}, TArg)" />
+    /// </summary>
+    private readonly T?[] _projects;
+
+    /// <summary>
+    /// Domain of ProjectIds for which values may be associated. Set equivalent to <see cref="_projectIdToIndex"/>'s keys.
+    /// Compared to input in <see cref="WithProjectIds(IReadOnlyList{ProjectId})"/> to determine if
+    /// the existing frozen dictionary can be reused.
+    /// </summary>
+    private readonly IReadOnlyList<ProjectId> _projectIds;
+
+    private ProjectIdMapping(IReadOnlyList<ProjectId> projectIds, FrozenDictionary<ProjectId, int> projectIdToIndex)
+    {
+        _projectIds = projectIds;
+        _projectIdToIndex = projectIdToIndex;
+        _projects = new T?[_projectIds.Count];
```

Comment:
curious if this ca be deferred.  i imaginet there are a bunch of intermediary solution instances that don't need these.

Reply:
That was the exact change I was working on this morning